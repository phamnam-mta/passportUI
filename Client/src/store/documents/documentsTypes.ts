export enum DocumentStatus {
    Loading = "Loading",
    Loaded = "Loaded",
    Analyzing = "Analyzing",
    Analyzed = "Analyzed",
    AnalyzeFailed = "AnalyzeFailed",
    Labeled = "Labeled",
}

export interface IDocumentStates {
    loadingStatus: DocumentStatus.Loaded | DocumentStatus.Loading;
    analyzingStatus?: DocumentStatus.Analyzed | DocumentStatus.Analyzing | DocumentStatus.AnalyzeFailed;
    labelingStatus?: DocumentStatus.Labeled;
}

export interface IRawDocument {
    name: string;
    type: string;
    url: string;
}

export interface IDocument extends IRawDocument {
    thumbnail: string;
    numPages: number;
    currentPage: number;
    states: IDocumentStates;
    expirationTime?: number;
}
