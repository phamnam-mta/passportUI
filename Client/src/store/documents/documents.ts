import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import { resetPredictions } from "store/predictions/predictions";
import { DocumentLoaderFactory, IDocumentLoader } from "utils/documentLoader";
import { DocumentStatus, IDocument, IRawDocument } from "./documentsTypes";

export type DocumentsState = { documents: IDocument[]; currentDocument: IDocument | null };

export const initialState: DocumentsState = { documents: [], currentDocument: null };

export const addDocuments = createAsyncThunk("documents/addDocuments", async (docsToAdd: IRawDocument[]) => {
    return Promise.all(docsToAdd.map(async (doc) => (await getLoader(doc)).loadDocumentMeta()));
});

export const setCurrentDocument = createAsyncThunk("documents/setCurrentDocument", async (document: IDocument) => {
    const loader = await getLoader(document);
    const documentPage = await loader.loadDocumentPage(document.currentPage);
    return { document, documentPage };
});

export const setCurrentPage = createAsyncThunk("documents/setCurrentPage", async (pageNumber: number, thunkApi) => {
    const { documents: state } = thunkApi.getState() as any;
    const loader = await getLoader(state.currentDocument);
    const documentPage = await loader.loadDocumentPage(pageNumber);
    return { pageNumber, documentPage };
});

const documentLoaders = new Map<string, IDocumentLoader>();

const getLoader = async (document: IRawDocument): Promise<IDocumentLoader> => {
    let loader = documentLoaders.get(document.url);
    if (!loader) {
        loader = await DocumentLoaderFactory.makeLoader(document);
        documentLoaders.set(document.url, loader);
    }

    return loader;
};

const documentsSlice = createSlice({
    name: "documents",
    initialState,
    reducers: {
        deleteDocument(state, action) {
            const docNameToDelete = action.payload.name;
            state.documents = state.documents.filter((document) => document.name !== docNameToDelete);
            documentLoaders.delete(action.payload.url);
        },
        setDocumentAnalyzingStatus(state, action) {
            const { name, status } = action.payload;
            const document = state.documents.find((document) => document.name === name);
            if (document) {
                document.states.analyzingStatus = status;
            }
            if (state.currentDocument?.name === name) {
                state.currentDocument!.states.analyzingStatus = status;
            }
        },
        setDocumentLabelingStatus(state, action) {
            const { name, status } = action.payload;
            const document = state.documents.find((document) => document.name === name);
            if (document) {
                document.states.labelingStatus = status;
            }
            if (state.currentDocument?.name === name) {
                state.currentDocument!.states.labelingStatus = status;
            }
        },
        clearCurrentDocument(state) {
            state.currentDocument = null;
        },
    },
    extraReducers: (builder) => {
        builder
            .addCase(resetPredictions, (state) => {
                state.documents.forEach((document) => {
                    document.states = { ...document.states, analyzingStatus: undefined };
                });
                if (state.currentDocument) {
                    state.currentDocument.states = { ...state.currentDocument.states, analyzingStatus: undefined };
                }
            })
            .addCase(addDocuments.pending, (state, action) => {
                const docs = action.meta.arg;
                docs.forEach((doc) => {
                    state.documents.push({
                        ...doc,
                        thumbnail: "",
                        numPages: 0,
                        currentPage: 0,
                        states: { loadingStatus: DocumentStatus.Loading },
                    });
                });
            })
            .addCase(setCurrentDocument.pending, (state, action) => {
                const document = action.meta.arg;
                const selectedDoc = state.documents.find((doc) => doc.name === document.name);
                if (selectedDoc) {
                    selectedDoc.states.loadingStatus = DocumentStatus.Loading;
                }
            })
            .addCase(setCurrentDocument.fulfilled, (state, action) => {
                const document: IDocument = {
                    ...action.payload.document,
                    states: {
                        ...action.payload.document.states,
                        loadingStatus: DocumentStatus.Loaded,
                    },
                };
                const selectedDoc = state.documents.find((doc) => doc.name === document.name);
                if (selectedDoc) {
                    selectedDoc.states.loadingStatus = DocumentStatus.Loaded;
                }
                state.currentDocument = document;
            })
            .addCase(setCurrentPage.fulfilled, (state, action) => {
                const { pageNumber } = action.payload;
                const document = state.documents.find((doc) => doc.name === state.currentDocument?.name);
                if (document) {
                    document.currentPage = pageNumber;
                }
                if (state.currentDocument) {
                    state.currentDocument.currentPage = pageNumber;
                }
            })
            .addCase(addDocuments.fulfilled, (state, action) => {
                const addedDocs = action.payload;
                const { documents } = state;
                addedDocs.forEach((doc) => {
                    const iDoc = documents.findIndex((d) => d.name === doc.name);
                    if (iDoc >= 0) {
                        documents[iDoc] = {
                            ...doc,
                            states: { ...documents[iDoc].states, loadingStatus: DocumentStatus.Loaded },
                        };
                    }
                });
            });
    },
});

export const { deleteDocument, setDocumentAnalyzingStatus, setDocumentLabelingStatus, clearCurrentDocument } =
    documentsSlice.actions;

export const reducer = documentsSlice.reducer;
