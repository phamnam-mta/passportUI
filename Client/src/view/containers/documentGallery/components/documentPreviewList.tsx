import * as React from "react";
import { IList, List, mergeStyleSets, ScrollToMode } from "@fluentui/react";

import { IDocument } from "store/documents/documentsTypes";
import DocumentPreview from "./documentPreview";

interface IDocumentPreviewListProps {
    currentDocument: IDocument | null;
    documents: Array<IDocument>;
    onDocumentClick: (name: string, index: number) => void;
    onDocumentDelete: (name: string, index: number) => void;
}

export class DocumentPreviewList extends React.PureComponent<IDocumentPreviewListProps> {
    private listRef: React.RefObject<IList>;
    private itemHeight = 124;

    constructor(props) {
        super(props);
        this.listRef = React.createRef();
    }

    public componentDidMount() {
        this.styleSets.listStyles.overflowY = "overlay";
        this.classNames = mergeStyleSets(this.styleSets);
    }

    public componentDidUpdate(prevProps) {
        const { currentDocument } = this.props;

        if (prevProps.currentDocument !== currentDocument && currentDocument) {
            this.scrollCurrentDocToTop();
        }
    }

    private styleSets = {
        focusZone: {
            width: "100%",
            height: "100%",
        },
        listStyles: {
            width: "100%",
            height: "100%",
            overflowY: "auto",
        },
        listItemStyles: {
            marginTop: 8,
        },
    };

    private classNames = mergeStyleSets(this.styleSets);

    public render() {
        return (
            <List
                componentRef={this.listRef}
                className={this.classNames.listStyles}
                items={this.props.documents}
                onRenderCell={this.onRenderCell}
            />
        );
    }

    private handleClick = (name: string, index: number) => {
        this.props.onDocumentClick(name, index);
    };

    private handleDocumentDelete = (name: string, index: number) => {
        this.props.onDocumentDelete(name, index);
    };

    private onRenderCell = (item: IDocument | undefined, index: number | undefined) => {
        return item ? (
            <div className={this.classNames.listItemStyles}>
                <DocumentPreview
                    selected={item.name === this.props.currentDocument?.name}
                    documentName={item.name}
                    documentStates={item.states}
                    documentImageSrc={item.thumbnail}
                    onDocumentClick={(name) => this.handleClick(name, Number(index))}
                    onDocumentDelete={(name) => this.handleDocumentDelete(name, Number(index))}
                />
            </div>
        ) : null;
    };

    private scrollCurrentDocToTop = () => {
        const { currentDocument, documents } = this.props;
        const index = documents.findIndex((doc) => doc.name === currentDocument!.name);
        if (index !== -1) {
            this.listRef.current?.scrollToIndex(index, () => this.itemHeight, ScrollToMode.auto);
        }
    };
}
