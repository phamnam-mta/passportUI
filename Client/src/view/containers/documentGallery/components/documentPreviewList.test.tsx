import { shallow } from "enzyme";
import * as React from "react";

import { DocumentStatus } from "store/documents/documentsTypes";
import { DocumentPreviewList } from "./documentPreviewList";

describe("<DocumentPreviewList />", () => {
    let baseProps;

    const items = new Array(10).fill({}).map((item, index) => {
        return {
            name: `item ${index}`,
            url: `/item_${index}.jpg`,
            status: index < 5 ? DocumentStatus.Loaded : DocumentStatus.Analyzing,
        };
    });

    const nullItems = new Array(10).fill(undefined);

    beforeEach(() => {
        baseProps = {
            currentDocument: null,
            documents: items,
            onDocumentClick: jest.fn(),
            onDocumentDelete: jest.fn(),
        };
    });

    describe("Rendering", () => {
        it("should match snapshot.", () => {
            const wrapper = shallow(<DocumentPreviewList {...baseProps} />);
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe("Event Handling", () => {
        it("should be empty list, when undefined items provided.", () => {
            const props = { ...baseProps, documents: nullItems };
            const wrapper = shallow(<DocumentPreviewList {...props} />);
            const list = wrapper.find("List") as any;

            expect(list.prop("onRenderCell")(nullItems[0], 0)).toBeNull();
        });

        it("should call onDocumentClick, when clicked on element.", () => {
            const clickedDoc = items[0];
            const wrapper = shallow(<DocumentPreviewList {...baseProps} />);

            const list = wrapper.find("List") as any;
            const documentPreview = shallow(list.prop("onRenderCell")(clickedDoc, 0)).find("DocumentPreview") as any;
            documentPreview.prop("onDocumentClick")(clickedDoc.name);

            expect(baseProps.onDocumentClick).toBeCalledWith(clickedDoc.name, 0);
        });

        it("should call onDocumentDelete, when clicked on element.", () => {
            const docIndexToDelete = 0;
            const docToDelete = items[docIndexToDelete];
            const wrapper = shallow(<DocumentPreviewList {...baseProps} />);

            const list = wrapper.find("List") as any;
            const documentPreview = shallow(list.prop("onRenderCell")(docToDelete, docIndexToDelete)).find(
                "DocumentPreview"
            ) as any;
            documentPreview.prop("onDocumentDelete")(docToDelete.name);

            expect(baseProps.onDocumentDelete).toBeCalledWith(docToDelete.name, docIndexToDelete);
        });

        it("should scroll to index, when currentDocument is changed", () => {
            const mockScrollToIndex = jest.fn();

            const wrapper = shallow(<DocumentPreviewList {...baseProps} />) as any;
            wrapper.instance().listRef = { current: { scrollToIndex: mockScrollToIndex } };
            wrapper.setProps({ currentDocument: items[0] });

            expect(mockScrollToIndex).toBeCalledTimes(1);
            expect(mockScrollToIndex.mock.calls[0][0]).toBe(0);
        });
    });
});
