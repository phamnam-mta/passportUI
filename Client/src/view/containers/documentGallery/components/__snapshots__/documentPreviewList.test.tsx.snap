// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DocumentPreviewList /> Rendering should match snapshot. 1`] = `
<List
  className="listStyles-110"
  componentRef={
    Object {
      "current": null,
    }
  }
  items={
    Array [
      Object {
        "name": "item 0",
        "status": "Loaded",
        "url": "/item_0.jpg",
      },
      Object {
        "name": "item 1",
        "status": "Loaded",
        "url": "/item_1.jpg",
      },
      Object {
        "name": "item 2",
        "status": "Loaded",
        "url": "/item_2.jpg",
      },
      Object {
        "name": "item 3",
        "status": "Loaded",
        "url": "/item_3.jpg",
      },
      Object {
        "name": "item 4",
        "status": "Loaded",
        "url": "/item_4.jpg",
      },
      Object {
        "name": "item 5",
        "status": "Analyzing",
        "url": "/item_5.jpg",
      },
      Object {
        "name": "item 6",
        "status": "Analyzing",
        "url": "/item_6.jpg",
      },
      Object {
        "name": "item 7",
        "status": "Analyzing",
        "url": "/item_7.jpg",
      },
      Object {
        "name": "item 8",
        "status": "Analyzing",
        "url": "/item_8.jpg",
      },
      Object {
        "name": "item 9",
        "status": "Analyzing",
        "url": "/item_9.jpg",
      },
    ]
  }
  onRenderCell={[Function]}
  renderedWindowsAhead={2}
  renderedWindowsBehind={2}
  startIndex={0}
/>
`;
